---
title: sequence matching - test data
author: Kai He
output: pdf_document
---


## Purpose: Match amino acid sequences identified by mass spectrometry to all 

#### dependencies: seqinr, dplyr, tidyr, ggplot2, "assertthat", "stringr"
```{r}
lapply(c("dplyr", "seqinr", "tidyr", "ggplot2","assertthat","stringr"), library, character.only = T)
```

#### default arguments for automated build
```{r}
lkup_path <- "../data/testData.csv" # use testdata
ref_path <- "../data/uniprot-all.fasta" # reference fasta file
```


read csv file from test data:
```{r}
ori_lkup <- read.csv(lkup_path)
# ori_lkup <- filter(ori_lkup, ST.2 != "index")
```

show head of data:
```{r}
head(ori_lkup)
```

select only relevant columns. the name "leading protein" is converted to "leading.protein" by R default

```{r}
lkup <- ori_lkup %>% 
  select(peptide
         # = X.5
         , leading.protein
         # = X.7
         ) %>% 
  # filter out proteins with no sequence information, doesn't matter for test but will become important for real data
  filter(peptide != "") %>% 
  
  # create new column indicating whether the N-terminus is blocked before tryptic digestion
  mutate(Ncapped = 0)

lkup$Ncapped[grepl("n",ignore.case = FALSE,lkup$peptide)] <- 1
```

variable that stores the peptide sequence in character

```{r}
pepseq <- as.character(lkup$peptide)

# show peptide sequences:

head(pepseq)
```

temp variable for formatting sequence
```{r}
tmp <- sub(".?\\.","",pepseq)
tmp <- sub("n\\[[0-9]*\\.?[0-9]*\\]","",tmp)
tmp <- gsub("\\[[0-9]*\\.?[0-9]*\\]","",tmp)
tmp <- sub("\\..*","",tmp)

# show tmp:
head(tmp)
```

Add the formatted sequence as a new column named Lseq

```{r}
# lead.protein <- sub(,"",lkup$Acc_id)

lkup <- lkup %>%
  mutate(Lseq=tmp) %>%
  rename(Acc_id = leading.protein) %>%
  filter(nchar(as.character(Acc_id))<=8) %>%
  mutate(Acc_id = sub("\\-[0-9]*","",Acc_id))

# write cleaned up lookup sequence to results folder
write.csv(lkup, "../results/lkup.csv")
```

show cleaned data:
```{r}
head(lkup)
```

Load and show the ref dataset.
```{r}
orig_ref <- read.fasta(ref_path, seqtype = "AA", as.string = T)

head(orig_ref)
```

clean the reference
```{r}
ref <- data.frame(Acc_id = getName(orig_ref), 
                  Rseq = rapply(getSequence(orig_ref, as.string = T), c)) %>%
  ### Clean Acc_id.
  mutate(Acc_id = substr(as.character(Acc_id), 4, 9)) %>%
  ### Clean Rseq.
  mutate(Rseq = toupper(Rseq)) %>%
  ### Remove duplicates.
  distinct(Acc_id, Rseq)
```

Write the cleaned reference to results folder
```{r}
write.csv(ref,"../results/ref.csv")
```

show cleaned ref:
```{r}
head(ref)
```


Merge the lkup and ref datasets by Acc_id.
```{r}
lkup <- read.csv("../results/lkup.csv")
ref <- read.csv("../results/ref.csv")

combo <- merge(lkup,ref, by = "Acc_id") %>% 
  ### Keep obs with unique values in Lseq and Rseq only.
  group_by(Rseq) %>%
  filter(!duplicated(Lseq)) 
  # select(Acc_id,Rseq,Lseq) %>% 
  # unique()

combo$X.x = NULL
combo$X.y = NULL
```

Check if all Acc_ids from lkup are in combo. Must be TRUE.
```{r}
assert_that(length(intersect(combo$Acc_id, lkup$Acc_id)) == length(unique(lkup$Acc_id))) 
```

Check if all Acc_ids from ref are in combo. Must be FALSE.
```{r}
assert_that(length(intersect(combo$Acc_id, as.character(ref$Acc_id))) != length(unique(ref$Acc_id)))
```

Set Match to 1 if Lseq is a substring of Rseq; 0 otherwise.
```{r}
combo$Lseq <- as.character(combo$Lseq)
combo$Rseq <- as.character(combo$Rseq)

```


combo data frame summarising matched Lseqs and Rseqs
```{r}
combo <- combo %>% 
  mutate(Match = mapply(grepl,Lseq,Rseq)) %>%
  ### Keep only matched obs.
  filter(Match == TRUE) %>% 
  ### Count character length of Lseq and Rseq.
  mutate(Llen = nchar(Lseq), 
         Rlen = nchar(Rseq),
         ### Get start and end indices of matched Lseq.
         Lstart = mapply(regexpr,Lseq,Rseq,fixed = TRUE),
         Lend = Lstart + Llen -1) %>% 
  ### Remove case with Lstart < 0.
  filter(Lstart>=0) %>% 
  ### Sort by Acc_id, Lstart and Lend.
  arrange(Acc_id,Lstart,Lend) 

```

Check if all obs are matched. Must be TRUE.
```{r}
assert_that(sum(combo$Match) == nrow(combo))
### See which obs are not matched.
which(combo$Match == FALSE)
```


Check if each Acc_id is associated with a unique Rseq. Must be TRUE.
```{r}
assert_that(length(unique(combo$Acc_id)) == length(unique(combo$Rseq)))
```

show combo:
```{r}
head(combo)
```

get unique reference sequences only.
```{r}
matchedRseq <- combo %>% 
  select(Acc_id,Rseq) %>% 
  distinct(.keep_all = TRUE)
```

show ref seqs:
```{r}
(matchedRseq)
```


create a vector of access ids repeated by length of corresponding reference sequences.
```{r}
accid <- sapply(as.integer(rownames(matchedRseq)),
                function(i){
                  replicate(nchar(matchedRseq$Rseq[i]),
                            expr = matchedRseq$Acc_id[i])})
```

show accid
```{r}
str(accid)
summary(accid)
```


store the level of accid in a tmp data frame df.
```{r}
df <- data.frame(ACID <- c(unlist(accid)))

df <- df %>% 
  # add Actual access id column
  mutate(Acc_id = as.factor(levels(unlist(accid))[ACID....c.unlist.accid..]),
         # add index for each accid
         index = rownames(df)) %>% 
  # arrange by Access id and index
  arrange(Acc_id,index) %>% 
  # group by access id.
  group_by(Acc_id) %>%
  # recreate index column, remove the level numbers, create new variable indicating matched or not
  mutate(index = row_number(),
         ACID....c.unlist.accid..=NULL) 

```

tmp vector storing start location of each Lseq
```{r}
# Lseq_starts <- (sapply(tdf$Acc_id, function(i) combo$Lstart[combo$Acc_id == i & combo$Ncapped == 0]))
Lseq_starts <- (sapply(df$Acc_id, function(i) combo$Lstart[combo$Acc_id == i & combo$Ncapped == 0]))
```

show Lseq_starts
```{r}
head(Lseq_starts)
```


new column storing first matching positions between Lseq and Rseq by getting first start location of matching Lseq
```{r}
df$firstmatch <- sapply(Lseq_starts,function(i) i[[1]][1])
```

divide the Rseq index into groups by first match of Lseq
```{r}
df$group <- replicate(nrow(df),0)
df$group[df$index > df$firstmatch-1] <- 1
```

find strat position of each possible Rseq marked by a K or R
```{r}
df$Rpep_start <- 0
```

get Rpep_index
```{r}
index_lst <- lapply(matchedRseq$Rseq,function(i) {
  which(df$Acc_id == as.character(matchedRseq$Acc_id[which(matchedRseq$Rseq == i)]) &
          df$index %in% unlist(gregexpr("[K,R]",i)))
})


df$Rpep_start[as.integer(unlist(index_lst))+1] <- 1

df <- df %>%   group_by(Acc_id) %>%   mutate(Rpep_index = sapply(index, function(i) sum(Rpep_start[1:i]))+1)
```

rename the tmp dataframe to grouped_peps, make new column stating if the peptide is matched
```{r}

grouped_peps <- df %>%
  group_by(Acc_id) %>%
  mutate(match = Rpep_index %in% Rpep_index[index %in% combo$Lstart[combo$Acc_id == Acc_id[1] & combo$Ncapped != 1]])

```

show grouped_peps:
```{r}
head(grouped_peps)
```

## basic statistics

### - % of amino acids matched [plot as scatterplot vs total number of amino acids]

assume no overlaps between Lseqs (will break Lseqs manually at possible cut sites)
% calculated as sum of Llen over length of corresponding Rseq.

```{r}
matchedRseq %>% 
  mutate(pct_matchaa = sum(combo$Llen[combo$Acc_id == Acc_id & combo$Ncapped == 0]) / combo$Rlen[combo$Acc_id == Acc_id][1] * 100,
         Rlen = combo$Rlen[combo$Acc_id == Acc_id][1]) %>% 
  ggplot(aes(x=Rlen, y = pct_matchaa)) + 
  geom_point()

```

### - % of peptides matched [plot as scatterplot vs total number of peptides]

```{r}
pctpep <- grouped_peps %>%
  group_by(Acc_id) %>%
  summarise(match_peps = nlevels(as.factor(Rpep_index[match])),
            total_peps = nlevels(as.factor(Rpep_index)),
            pct_matchpep = match_peps / total_peps * 100)

ggplot(pctpep,aes(x = total_peps, y = pct_matchpep)) + geom_point()
```

### - % of peptides with identification probability =1 matched [plot as scatterplot vs total number of peptides]

```{r}
pctlongpep <- grouped_peps %>% 
  group_by(Acc_id,Rpep_index) %>% 
  summarise(peplen = n(),matched = any(match)) %>% 
  group_by(Acc_id) %>% 
  summarise(total_peps = n(),
            long_peps = sum(peplen>6),
            matched_long_peps = sum(peplen>6 & matched),
            pct_longmatch = matched_long_peps / long_peps * 100,
            pct_longpep = long_peps / total_peps * 100)

ggplot(pctlongpep,aes(x= pct_longmatch)) + geom_histogram()

```

### - start position of first match (both as Nth amino acid and Nth peptide) (old GAPS I believe) [plot as histogram]

```{r}
first_match_start <- grouped_peps %>% 
  group_by(Acc_id) %>% 
  summarise(firstmatch_aa = min(firstmatch),
            firstmatch_pep = min(Rpep_index[group == 1]),
            percentile_aa = firstmatch_aa / n(),
            percentile_pep = (firstmatch_pep - 1) / max(Rpep_index))

p <- ggplot(first_match_start)
p+  geom_histogram(aes(x=firstmatch_aa, alpha = 0.5, fill = "aa"),binwidth = 50)
p+  geom_histogram(aes(x=firstmatch_pep, alpha = 0.5, fill = "pep"),binwidth = 10)
```


### - distance of endposition of last match (both as Nth amino acid and Nth peptide) for resed end (old GAPE I believe) [plot as histogram]

```{r}
combo %>% 
  group_by(Acc_id) %>% 
  summarise(lastmatch_aa = max(Lend),
            dist_aa = Rlen[1] - lastmatch_aa) %>% 
  ggplot(aes(x=dist_aa)) + geom_histogram(binwidth = 10)


grouped_peps %>% 
  group_by(Acc_id) %>% 
  summarise(lastmatch_pep = max(Rpep_index[match]),
            dist_pep = max(Rpep_index) - lastmatch_pep) %>% 
  ggplot(aes(x=dist_pep)) + geom_histogram(binwidth = 5)
```

### - # and percentage of inner gaps (both in terms of amino acids and #peptides) [plot as histogram]

```{r}
pgap <- grouped_peps %>% 
  group_by(Acc_id) %>% 
  summarise(n_gap_aa = max(index[match])-min(index[match]) - sum(match) + 1,
            total_aa = n(),
            n_gap_pep = sum(diff(match) > 0 ) - 1,
            total_pep = nlevels(as.factor(Rpep_index))) %>% 
  ggplot()


pgap + geom_histogram(aes(x = n_gap_aa),binwidth = 10)
pgap + geom_freqpoly(aes(x = n_gap_aa / total_aa * 100))
pgap + geom_histogram(aes(x = n_gap_pep),binwidth = 1)
pgap + geom_freqpoly(aes(x = n_gap_pep / total_pep * 100))
```

###   - Hypergeometric test of peptide distribution between the two groups (start to before first match, first match to end) as discussed [plot as scatter vs some of the above values, e.g. %peptides matched, start position of first match, …]

plot the probability that the observed first match position is due to random chance, against the first match position as a percentile in the protein sequence.
```{r}
grouped_peps %>% 
  group_by(Acc_id) %>% 
  summarise(q = nlevels(as.factor(Rpep_index[match])), # number of peptides matched
            m = nlevels(as.factor(Rpep_index[group==1])), # number of peptides in group 1
            n = nlevels(as.factor(Rpep_index[group==0])), # number of peptides in group 0
            k = nlevels(as.factor(Rpep_index[match])), # numbero of peptides matched
            pr = phyper(q,m,n,k) - phyper(q-1,m,n,k)) %>% 
  full_join(first_match_start) %>% 
  ggplot(aes(x = percentile_aa, y = pr))+
  geom_point()
```

